File: app/src/main/java/ru/mamykin/foreignbooksreader/common/Paginator.java
package ru.mamykin.foreignbooksreader.common;

import android.support.annotation.Nullable;
import android.text.Layout;
import android.text.StaticLayout;
import android.text.TextPaint;

import java.util.ArrayList;
import java.util.List;

/**
 * Creation date: 5/29/2017
 * Creation time: 11:39 AM
 * @author Andrey Mamykin(mamykin_av)
 */
public class Paginator {
    private final List<CharSequence> pagesList = new ArrayList<>();
    private int currentPageIndex = 0;

    public Paginator(CharSequence content, int width, int height, TextPaint paint,
                     float spacingMult, float spacingAdd, boolean includePad) {

        final StaticLayout layout = new StaticLayout(content, paint, width,
                Layout.Alignment.ALIGN_NORMAL, spacingMult, spacingAdd, includePad);

        final int lines = layout.getLineCount();
        final CharSequence text = layout.getText();
        int startOffset = 0;
        int lHeight = height;

        for (int i = 0; i < lines; i++) {
            if (lHeight < layout.getLineBottom(i)) {
                addPage(text.subSequence(startOffset, layout.getLineStart(i)));
                startOffset = layout.getLineStart(i);
                lHeight = layout.getLineTop(i) + height;
            }
            if (i == lines - 1) {
                addPage(text.subSequence(startOffset, layout.getLineEnd(i)));
                return;
            }
        }
    }

    private void addPage(CharSequence text) {
        pagesList.add(text);
    }

    public int getPagesCount() {
        return pagesList.size();
    }

    @Nullable
    public CharSequence getPage(int index) {
        return (index >= 0 && index < pagesList.size()) ? pagesList.get(index) : null;
    }

    public CharSequence getCurrentPage() {
        return getPage(currentPageIndex);
    }

    public float getReadPercent() {
        return (((getCurrentIndex() + 1) / (float) getPagesCount()) * 100);
    }

    public String getReadPages() {
        return (getCurrentIndex() + 1) + "/" + getPagesCount();
    }

    public int getCurrentIndex() {
        return currentPageIndex;
    }

    public void setCurrentIndex(int currentPageIndex) {
        this.currentPageIndex = currentPageIndex;
    }
}

File: app/src/main/java/ru/mamykin/foreignbooksreader/domain/readbook/Paginator.kt
package ru.mamykin.foreignbooksreader.domain.readbook

import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint

import java.util.ArrayList

class Paginator(content: CharSequence,
                width: Int,
                height: Int,
                paint: TextPaint,
                spacingMult: Float,
                spacingAdd: Float,
                includePad: Boolean
) {

    private val pagesList = ArrayList<CharSequence>()
    var currentIndex = 0

    val pagesCount: Int
        get() = pagesList.size

    val currentPage: CharSequence?
        get() = getPage(currentIndex)

    val readPercent: Float
        get() = (currentIndex + 1) / pagesCount.toFloat() * 100

    val readPages: String
        get() = (currentIndex + 1).toString() + "/" + pagesCount

    init {
        val layout = StaticLayout(content, paint, width,
                Layout.Alignment.ALIGN_NORMAL, spacingMult, spacingAdd, includePad)

        val lines = layout.lineCount
        val text = layout.text
        var startOffset = 0
        var lHeight = height

        for (i in 0 until lines) {
            if (lHeight < layout.getLineBottom(i)) {
                addPage(text.subSequence(startOffset, layout.getLineStart(i)))
                startOffset = layout.getLineStart(i)
                lHeight = layout.getLineTop(i) + height
            }
            if (i == lines - 1) {
                addPage(text.subSequence(startOffset, layout.getLineEnd(i)))
                break
            }
        }
    }

    private fun addPage(text: CharSequence) {
        pagesList.add(text)
    }

    fun getPage(index: Int): CharSequence? {
        return if (index >= 0 && index < pagesList.size) pagesList[index] else null
    }
}